core/src/main/java/org/infinispan/marshall/protostream/impl/MarshallableArray.java

```diff
@@ -0,0 +1,173 @@
+ * <code>Collection<MarshallableObject></code> without the overhead of creating a {@link MarshallableObject} per
+ * entry.
+ *
+ * @author Ryan Emerson
+ * @since 16.0
+ */
+@ProtoTypeId(ProtoStreamTypeIds.MARSHALLABLE_ARRAY)
+public class MarshallableArray<T> {
+
+   static final Log log = LogFactory.getLog(MarshallableArray.class);
+
+   /**
+    * @param entries an Array to be wrapped as a {@link MarshallableArray}.
+    * @return a new {@link MarshallableArray} instance containing the passed object if the array is not null, otherwise
+    * null.
+    */
+   public static <T> MarshallableArray<T> create(T[] entries) {
+      return entries == null ? null : new MarshallableArray<>(entries);
+   }
+
+   /**
+    * @param wrapper the {@link MarshallableArray} instance to unwrap.
+    * @return the wrapped {@link Collection} or null if the provided wrapper does not exist.
+    */
+   public static <T> T[] unwrap(MarshallableArray<T> wrapper, T[] array) {
+      if (wrapper == null)
+         return null;
+
+      if (wrapper.array != null)
+         return wrapper.array;
+
+      assert wrapper.collection != null;
+      wrapper.array = wrapper.collection.toArray(array);
+      return wrapper.array;
+   }
+
+   private volatile T[] array;
+   private volatile Collection<T> collection;
+
+   private MarshallableArray(T[] array) {
+      this.array = array;
+   }
+
+   private MarshallableArray(Collection<T> collection) {
+      this.collection = collection;
+   }
+
+   @ProtoFactory
+   MarshallableArray(int size, List<byte[]> bytes) {
+      throw log.marshallerNotOverridden(getClass().getName());
+   }
+
+   @ProtoField(1)
+   int size() {
+      throw log.marshallerNotOverridden(getClass().getName());
+   }
+
+   @ProtoField(number = 2, collectionImplementation = ArrayList.class)
+   List<byte[]> getBytes() {
+      throw log.marshallerNotOverridden(getClass().getName());
+   }
+
+   public Object[] get() {
+      return array != null ? array : collection.toArray();
+   }
+
+   public static class Marshaller extends GeneratedMarshallerBase implements ProtobufTagMarshaller<MarshallableArray> {
+      private final String typeName;
+      private final org.infinispan.commons.marshall.Marshaller marshaller;
+
+      public Marshaller(org.infinispan.commons.marshall.Marshaller marshaller) {
+         this.typeName = getFqTypeName(MarshallableArray.class);
+         this.marshaller = marshaller;
+      }
+
+      @Override
+      public MarshallableArray read(ReadContext ctx) throws IOException {
+         final TagReader in = ctx.getReader();
+         try {
+            int tag = in.readTag();
+            if (tag == 0)
+               return new MarshallableArray<>(List.of());
+
+            if (tag != (1 << WireType.TAG_TYPE_NUM_BITS | WireType.WIRETYPE_VARINT))
+               throw new IllegalStateException("Unexpected tag: " + tag);
+
+            final int size = in.readInt32();
+            List<Object> entries = new ArrayList<>(size);
```

Comment:
why not `Object[] entries = new Object[size];`? You need to cast to `T[]` but avoid any copying.